# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Load Block File
#'
#' @description
#' Loads a genetic block file, typically containing linkage disequilibrium (LD) block definitions.
#'
#' @details
#' This function interfaces with C++ code to read and process block files used in genomic analyses,
#' such as defining independent regions for summary statistics processing.
#'
#' @param block_file Character string specifying the path to the block file.
#'
#' @return The result from the underlying C++ function, likely a list or data frame of block information.
#'
#' @examples
#' \dontrun{
#' blocks <- load_block_file("path/to/block_file.txt")
#' }
#'
#' @seealso \code{\link{test_blocks}}, \code{\link{Cal_blockinf}}
#'
#' @export
load_block_file <- function(block_file) {
    .Call(`_MERLIN_load_block_file`, block_file)
}

#' Test Blocks
#'
#' @description
#' Assigns or tests genetic positions (base pairs and chromosomes) against predefined blocks.
#'
#' @details
#' Used to map SNPs to LD blocks for downstream analyses like clumping or independence testing.
#'
#' @param bp Numeric vector of base pair positions.
#' @param chr Numeric or character vector of chromosome identifiers.
#' @param block_file Character string path to the block file.
#'
#' @return Results of block assignment or testing, possibly indices or block IDs.
#'
#' @examples
#' \dontrun{
#' test_result <- test_blocks(bp = c(1000, 2000), chr = c(1, 1), block_file = "blocks.txt")
#' }
#'
#' @seealso \code{\link{load_block_file}}, \code{\link{Cal_blockR}}
#'
#' @export
test_blocks <- function(bp, chr, block_file) {
    .Call(`_MERLIN_test_blocks`, bp, chr, block_file)
}

#' Standard Set Difference
#'
#' @description
#' Computes the set difference between two vectors, removing elements of y from x.
#'
#' @details
#' A standardized implementation, possibly handling numeric or character vectors efficiently.
#'
#' @param x Vector from which to remove elements.
#' @param y Vector of elements to remove.
#'
#' @return Vector containing elements in x but not in y.
#'
#' @examples
#' diff <- std_setdiff(1:10, 5:15)
#'
#' @export
std_setdiff <- function(x, y) {
    .Call(`_MERLIN_std_setdiff`, x, y)
}

LDclump <- function(R, ld_r2_thresh) {
    .Call(`_MERLIN_LDclump`, R, ld_r2_thresh)
}

Cal_blockR <- function(bp, chr, avbIndex, idx4panel, block_file, stringname3, ld_r2_thresh, coreNum, lam) {
    .Call(`_MERLIN_Cal_blockR`, bp, chr, avbIndex, idx4panel, block_file, stringname3, ld_r2_thresh, coreNum, lam)
}

Cal_block_Rmatrix <- function(bp, chr, avbIndex, idx4panel, block_file, stringname3, ld_r2_thresh, coreNum, lam) {
    .Call(`_MERLIN_Cal_block_Rmatrix`, bp, chr, avbIndex, idx4panel, block_file, stringname3, ld_r2_thresh, coreNum, lam)
}

Cal_blockinf <- function(bp, chr, block_file) {
    .Call(`_MERLIN_Cal_blockinf`, bp, chr, block_file)
}

IndepSummary <- function(bp, chr, avbIndex, block_file, stringname3, bh1, bh2, se1, se2, coreNum, lam, ld_r2_thresh) {
    .Call(`_MERLIN_IndepSummary`, bp, chr, avbIndex, block_file, stringname3, bh1, bh2, se1, se2, coreNum, lam, ld_r2_thresh)
}

testR <- function(rho, n) {
    .Call(`_MERLIN_testR`, rho, n)
}

comb <- function(p) {
    .Call(`_MERLIN_comb`, p)
}

Mat2Vec <- function(R) {
    .Call(`_MERLIN_Mat2Vec`, R)
}

Vec2Mat <- function(RV, p1) {
    .Call(`_MERLIN_Vec2Mat`, RV, p1)
}

Cal_block_Rvec <- function(bp, chr, avbIndex, block_file, stringname3, coreNum, lam) {
    .Call(`_MERLIN_Cal_block_Rvec`, bp, chr, avbIndex, block_file, stringname3, coreNum, lam)
}

MRGEI_Gam3seo <- function(gammah1, gammah3, Gammah1, Gammah3, se1, se2, se3, se4, R, rho_1, rho_2) {
    .Call(`_MERLIN_MRGEI_Gam3seo`, gammah1, gammah3, Gammah1, Gammah3, se1, se2, se3, se4, R, rho_1, rho_2)
}

getLineNum <- function(filename) {
    .Call(`_MERLIN_getLineNum`, filename)
}

ReadSNPinfo <- function(stringname, A1, A2, rsname, chr, bp, morgan, N) {
    .Call(`_MERLIN_ReadSNPinfo`, stringname, A1, A2, rsname, chr, bp, morgan, N)
}

Read_summarystat <- function(stringname, SA1, SA2, rsname, betah, s2, pvalue, chr, bp, N) {
    invisible(.Call(`_MERLIN_Read_summarystat`, stringname, SA1, SA2, rsname, betah, s2, pvalue, chr, bp, N))
}

select <- function(vec_, idx_) {
    .Call(`_MERLIN_select`, vec_, idx_)
}

matchsnp <- function(stringname1, stringname2, stringname3, matchExp) {
    .Call(`_MERLIN_matchsnp`, stringname1, stringname2, stringname3, matchExp)
}

matchscreen <- function(screenname, stringname1, stringname2, stringname3, pva_cutoff, matchExp = FALSE) {
    .Call(`_MERLIN_matchscreen`, screenname, stringname1, stringname2, stringname3, pva_cutoff, matchExp)
}

phi <- function(x) {
    .Call(`_MERLIN_phi`, x)
}

cdfNormal <- function(x, mean, sd) {
    .Call(`_MERLIN_cdfNormal`, x, mean, sd)
}

MulticdfNormal <- function(x) {
    .Call(`_MERLIN_MulticdfNormal`, x)
}

multiphi <- function(x) {
    .Call(`_MERLIN_multiphi`, x)
}

inverseNormal <- function(prob, mean, sd) {
    .Call(`_MERLIN_inverseNormal`, prob, mean, sd)
}

MultiinverseNormal <- function(prob) {
    .Call(`_MERLIN_MultiinverseNormal`, prob)
}

MatSum <- function(y1, y2) {
    .Call(`_MERLIN_MatSum`, y1, y2)
}

truncEstfun <- function(a, b, x1, x2, maxIter, burnin, thin) {
    .Call(`_MERLIN_truncEstfun`, a, b, x1, x2, maxIter, burnin, thin)
}

